Plugins: Pass through dashboard/contextual HTTP headers to plugins/datasources

GiedriusS opened the pull request:
`X-Dashboard-Uid` and `X-Panel-Id` are two very useful headers set by Grafana front-end that we would like to see on the data source as well. This is so that it would be possible to pinpoint from where slow queries are coming in Mimir/Thanos/Cortex/etc., for example. Relevant Mimir code lines:
https://github.com/grafana/mimir/blob/0a94f26203cae7c1784a6d3040ccddb6c2063de9/pkg/frontend/transport/handler.go#L182-L184

Tested manually that with these changes the headers are visible.

GiedriusS committed:
services/query: pass through useful headers to data source

`X-Dashboard-Uid`, `X-Datasource-Uid`, `X-Grafana-Org-Id`, `X-Panel-Id` are very useful headers set
by Grafana front-end that we would like to see on the data source as
well. This is so that it would be possible to pinpoint from where slow
queries are coming in Mimir/Thanos/Cortex/etc., for example. Relevant
Mimir code lines:
https://github.com/grafana/mimir/blob/0a94f26203cae7c1784a6d3040ccddb6c2063de9/pkg/frontend/transport/handler.go#L182-L184

Tested manually that with these changes the headers are visible.

GiedriusS commented:
@sakjur @idafurjes @ying-jeanne ping :hugs:

marefr reviewed (changes requested):
Thank you for your contribution. This request make sense, but would like to see some changes since we recently refactored quite a lot of the header handling for plugins and in addition to make sure it applies to any plugin request, if headers are applicable/set on incoming HTTP request.

Can you please 
- create a new plugin client middleware, similar to https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/clientmiddleware/user_header_middleware.go, named HTTP header middleware or something and move your current implementation there.
- Use `SetHTTPHeader` to set headers on backend plugin requests, e.g. https://github.com/grafana/grafana/blob/4df78cebc2f007524b047f48a4b90f493a045350/pkg/services/pluginsintegration/clientmiddleware/user_header_middleware.go#L39
- Add tests
- Consider forwarding X-Grafana-Org-Id as well while you're at it based on  https://github.com/grafana/grafana/issues/41623#issuecomment-1309605378

GiedriusS commented:
> 

Thanks for your review! I have done all of the changes you requested except for (2). I think it's better to modify headers directly because:

- Frontend uses non-canonicalized header names so it would be less confusing to see the same on the data source;
- Prometheus data source accesses headers using `Headers` directly so with `SetHTTPHeader`  the data source would see headers such as `Http_x-Dashboard-Uid`. That's what I've seen after using `SetHTTPHeader`. Changing Prometheus data source to use `GetHTTPHeaders` is a separate breaking change IMHO so I have added the headers directly. I think this kind of cleanup would be much better suited for a separate PR.

I had more reasons but I forgot :smile: let me know what you think.

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware.go:
```diff
@@ -0,0 +1,94 @@
+		for _, httpHeaderName := range headersList {
+			t.Headers[httpHeaderName] = reqCtx.Req.Header.Get(httpHeaderName)
+		}
+	case *backend.CallResourceRequest:
```
You can skip this for `backend.CallResourceRequest` since most of incoming HTTP headers are already set.

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware.go:
```diff
@@ -0,0 +1,94 @@
+	switch t := req.(type) {
+	case *backend.QueryDataRequest:
+		for _, httpHeaderName := range headersList {
+			t.Headers[httpHeaderName] = reqCtx.Req.Header.Get(httpHeaderName)
```
Shouldn't populate the header if it's not set in incoming request. Given my explanation of the PR I'm working on I think SetHTTPHeader should be used since this is new feature for backend plugins and `backend.QueryDataRequest.Headers` is not entirely for HTTP headers - it's for environment/metadata about the reuqest.

Same applies for CheckHealthRequest below.

marefr reviewed (changes requested):
Thanks. 

> * Prometheus data source accesses headers using `Headers` directly so with `SetHTTPHeader`  the data source would see headers such as `Http_x-Dashboard-Uid`. That's what I've seen after using `SetHTTPHeader`. Changing Prometheus data source to use `GetHTTPHeaders` is a separate breaking change IMHO so I have added the headers directly. I think this kind of cleanup would be much better suited for a separate PR.

Sorry, forgot to tell you that I have PR #60417 in progress soon to be merged that removes Prometheus usage of forwarding the `Headers` inside of the datasource, it happens via middleware instead. I'll let you know when that is merged.

marefr commented:
https://github.com/grafana/grafana/pull/60417 is merged now. Please sync with the main branch.

GiedriusS commented:
Thx for https://github.com/grafana/grafana/pull/60417, I've rebased my PR on top of your changes. Everything works well now with the helpers SetHTTPHeader() and others. Double checked that the HTTP server sees the headers from a sample dashboard:

```
Path: /api/v1/query_range
Headers:
Host: localhost:8000
User-Agent: Grafana/9.4.0-pre
Content-Length: 57
Content-Type: application/x-www-form-urlencoded
X-Dashboard-Uid: lN53lOcVk
X-Datasource-Uid: aIyC_OcVz
X-Grafana-Org-Id: 1
X-Panel-Id: 2
Accept-Encoding: gzip
```

I've also added a check to not set the header if it is an empty string + removed CallResourceRequest from the headers modification.

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware.go:
```diff
@@ -0,0 +1,83 @@
+		req.SetHTTPHeader(headerName, gotVal)
+	}
+
+	return ctx
```
Nit. You can remove returning ctx since there's no changes to ctx

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware_test.go:
```diff
@@ -0,0 +1,68 @@
+			DataSourceInstanceSettings: &backend.DataSourceInstanceSettings{},
+		}
+
+		require.NoError(t, err)
```
Please move this line up directly after the error is set

marefr commented on pkg/services/pluginsintegration/clientmiddleware/tracing_header_middleware_test.go:
```diff
@@ -0,0 +1,68 @@
+			})
+			require.NoError(t, err)
+
+			require.Equal(t, `lN53lOcVk`, cdt.QueryDataReq.GetHTTPHeader(`X-Dashboard-Uid`))
```
Some test coverage if the headers are not set would be nice

marefr reviewed (approved):
LGTM nice. Some minor things left before we can merge, see comments.

@sakjur would you kindly help out getting this merged since I'm OOO until mid-January. Thank you 🙏

sakjur commented:
👍 willdo!

GiedriusS commented:
Thanks for the prompt reviews :bow: I have updated the code:

- Moved err check to just after the line where it is introduced in the test;
- Removed unused params from the function and the unnecessary return value;
- Added tests for unset/empty tracing headers.

sakjur reviewed (approved):
Looks good to me 👍

sakjur commented:
Thank you for your contribution, we'll include this in v9.4.0 😄
